### PR TITLE
Share url + lang

### LINKED ISSUE #495: Sharing URL + Language
Ex. say I want to share a URL that refers to a section that is only present in the java version of some module.

### PR DESCRIPTION
**Pull request by bqi343**
closes #495

### CODE REVIEW

**Comment by thecodingwizard** [comment]
looks like the query param isn't being persisted when the URL changes -- a quick hack could be to use https://www.gatsbyjs.com/docs/reference/config-files/gatsby-browser/#onRouteUpdate

This would cause even non-module/solution routes to have the query param though :/

**Comment by bqi343** [comment]
When I'm logged in, everything has a query param.

> Or maybe the query param should only affect that immediate page, and clicking a different module link would revert the language back to the user setting (this seems reasonable/easier?)

seems reasonable

File: gatsby-browser.tsx [hunk 1]
```diff
@@ -59,3 +59,8 @@ export const onClientEntry = () => {
     event.preventDefault();
   });
 };
+
+export const onRouteUpdate = ({ location, prevLocation }) => {
```
Comment (bqi343) [inline gatsby-browser.tsx:63]: @thecodingwizard is this supposed to print anything to console?

File: src/components/MarkdownLayout/MarkdownLayout.tsx [hunk 2]
```diff
@@ -64,9 +64,15 @@ export default function MarkdownLayout({
...
-  );
+  const {
+    userProgressOnModules,
+    setModuleProgress,
+    setLang,
+    lang,
+  } = useContext(UserDataContext);
+  React.useEffect(() => {
+    setLang(lang);
+  }, []);
```
Comment (bqi343) [inline src/components/MarkdownLayout/MarkdownLayout.tsx:75]: oops when the same module is clicked twice the lang disappears D:

File: gatsby-browser.tsx (hunk 1 above)
Comment (thecodingwizard) [reply gatsby-browser.tsx:63]: yes, though you might need to restart dev server (?)

**Comment by thecodingwizard** [comment]
remaining issues:

- clicking a module link twice causes the lang parameter to disappear (let's just pretend this is a feature...)
- multiple tabs open at the same time can cause issues with firebase synchronization (let's just pretend this bug doesn't exist; if it does happen it's just a minor inconvenience to switch the language anyway)

**Review by thecodingwizard — APPROVED** [review]
this is probably ready to merge, though I would probably test this a bit more ://

**Comment by bqi343** [comment]
> clicking a module link twice causes the lang parameter to disappear (let's just pretend this is a feature...)

@thecodingwizard do you know how to fix this?

**Comment by thecodingwizard** [comment]
Possibly by setting the language on every rerender (but idk how that will affect performance). But I think it can be considered a feature -- it provides a somewhat simple way to get rid of the lang query param in case the user wants to share the module link only without sharing the language.

**Pull request merged**